File: opensrp-app/src/main/java/org/smartregister/workmanager/utils/WorkerUtils.java
package org.smartregister.workmanager.utils;

import android.annotation.SuppressLint;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.os.Build;

import androidx.core.app.NotificationCompat;

import org.smartregister.R;
import org.smartregister.workmanager.Constants;

public final class WorkerUtils {

    public static void makeStatusNotification(Context context, int notificationId, String title, String message) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            @SuppressLint("WrongConstant") NotificationChannel channel = new NotificationChannel(Constants.CHANNEL_ID, Constants.CHANNEL_NAME, NotificationManager.IMPORTANCE_LOW);
            channel.setDescription(Constants.CHANNEL_DESC);

            notificationManager.createNotificationChannel(channel);
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, Constants.CHANNEL_ID)
                .setContentTitle(title)
                .setContentText(message)
                .setSmallIcon(R.drawable.ic_opensrp_logo)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setVibrate(new long[0]);

        notificationManager.notify(notificationId, builder.build());
    }
}

File: opensrp-core/src/main/java/org/smartregister/util/WorkerUtils.kt
package org.smartregister.util

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import org.smartregister.R
import kotlin.random.Random


object WorkerUtils {
    private const val CHANNEL_ID = "org.smartregisterx"
    private const val CHANNEL_NAME = "OpenSRP"
    private const val CHANNEL_DESC = "OpenSRP Client"

    fun makeStatusNotification(
        context: Context,
        notificationId: Int,
        title: String?,
        message: String?
    ) {
        val notificationManager = NotificationManagerCompat.from(context)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_LOW
            )
            channel.description = CHANNEL_DESC
            notificationManager.createNotificationChannel(channel)
        }
        val builder: NotificationCompat.Builder = NotificationCompat.Builder(context, CHANNEL_ID)
            .setContentTitle(title)
            .setContentText(message)
            .setSmallIcon(R.drawable.ic_opensrp_logo)
            .setPriority(NotificationCompat.PRIORITY_LOW)
        notificationManager.notify(notificationId, builder.build())
    }

    fun dismissNotification(context: Context, notificationId: Int){
        NotificationManagerCompat.from(context).cancel(notificationId)
    }
}

class WorkerNotificationDelegate(private val context: Context, private val title: String?) {

    private val notificationId: Int = 100

    fun notify(message: String) {
        WorkerUtils.makeStatusNotification(context, notificationId, title, message)
    }

    fun dismiss() {
        WorkerUtils.dismissNotification(context, notificationId)
    }
}